### PR TITLE
Enable pipelined compilation by default

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit enables pipelined compilation by default in Cargo now that
the requisite support has been stablized in rust-lang/rust#62766. This
involved minor updates in a number of locations here and there, but
nothing of meat has changed from the original implementation (just
tweaks to how rustc is called).

### CODE REVIEW

**Comment by alexcrichton** [comment]
r? @ehuss 

(note that CI will fail and this can't land until https://github.com/rust-lang/rust/pull/62766 propagates to nightly)

**Comment by alexcrichton** [comment]
Ok https://github.com/rust-lang/rust/pull/62766 has landed, I've rebased this and updated all failing tests. This will fail CI since the compiler isn't on the nightly channel yet, but @ehuss if you're up for it this should be good for review!

**Comment by ehuss** [comment]
I restarted CI, and it looks like stable/beta does not pass.

**Comment by alexcrichton** [comment]
Ah good point, I've updated the PR to just ignore those tests on non-nightly builders for now. Both of their assertions are tweaked when pipelining is enabled (both tests still morally pass on all compilers, it's just a difference of what should be asserted) and looking into the tests it seems not worth it to keep them running on stable/beta. Once pipelining rides to stable and we can remove the feature detection for that rustc flag we can update these tests to run on all channels again.

**Comment by ehuss** [comment]
r=me with minor nit addressed.

File: tests/testsuite/dep_info.rs [hunk 1]
```diff
@@ -511,6 +511,6 @@ fn canonical_path() {
     assert_deps_contains(
         &p,
         "target/debug/.fingerprint/foo-*/dep-lib-foo-*",
-        &[(1, "src/lib.rs"), (2, "debug/deps/libregdep-*.rlib")],
+        &[(1, "src/lib.rs"), (2, "debug/deps/libregdep-*.r*")],
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply tests/testsuite/dep_info.rs]: Sure yeah, the `r*` is to handle `rlib` on stable/beta but `rmeta` on nightly. I'll just switch this to a nightly-only test though and make it `rmeta`

**Pull request merged**